text/3185-static-async-fn-in-trait.md

```diff
@@ -0,0 +1,258 @@
+
+[async-trait]: https://github.com/dtolnay/async-trait
+[performance implications]: https://rust-lang.github.io/wg-async-foundations/vision/submitted_stories/status_quo/barbara_benchmarks_async_trait.html
+
+## Will anyone use async-trait crate once this RFC lands?
+
+The async-trait crate will continue to be useful after this RFC, because it allows traits to remain `dyn`-safe. This is a limitation in the current design that we plan to address in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+## The `async-trait` crate
+
+The most common way to use `async fn` in traits is to use the [`async-trait`] crate. This crate takes a different approach to the one described in this RFC. Async functions are converted into ordinary trait functions that return `Box<dyn Future>` rather than using an associated type. This means that the resulting traits are dyn safe and avoids a dependency on generic associated types, but it also has two downsides:
+
+* Requires a box allocation on every trait function call; while this is often no big deal, it can be prohibitive for some applications.
+* Requires the trait to state up front whether the resulting futures are `Send` or not. The [`async-trait`] crate defaults to `Send` and users write `#[async_trait(?Send)]` to disable this default.
+
+Since the async function support in this RFC means that traits are not dyn safe, we do not expect it to completely displace uses of the `#[async_trait]` crate.
+
+[`async-trait`]: https://crates.io/crates/async-trait
+
+## The real-async-trait crate
+
+The [`real-async-trait`] lowers `async fn` to use GATs and impl Trait, roughly as described in this RFC.
+
+[`real-async-trait`]: https://crates.io/crates/real-async-trait
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Dyn compatibility
+
+It is not a breaking change for traits to become dyn safe. We expect to make traits with async functions dyn safe, but doing so requires overcoming a number of interesting challenges, as described in the [async fundamentals evaluation doc][eval].
+
+## Impl trait in traits
+
+The [impl trait initiative] is expecting to propose "impl trait in traits" (see the [explainer](https://rust-lang.github.io/impl-trait-initiative/explainer/rpit_trait.html) for a brief summary). This RFC is compatible with the proposed design.
```

Comment:
this is still compatible with #3193 right?

Reply:
yes, the RFCs were developed somewhat in tandem, and they should be compatible.